### PR TITLE
Files/3

### PR DESCRIPTION
**Pull request by shdpl**
[upload] complete implementation
[testing,file-service] make use of disk instead of Storage because of lacking access to metadata api
[list] changed endpoint definition to use GET method and multipart-form parameters instead of POST JSON

### CODE REVIEW

**Review by qunabu — APPROVED** [review]

File: src/Http/Exceptions/Handler.php [hunk 1]
```diff
@@ -2,18 +2,20 @@
...
-     * @param  \Throwable  $e
-     * @return \Symfony\Component\HttpFoundation\Response
+     * @param Request $request
+     * @param Throwable $exception
+     * @return Response
      *
-     * @throws \Throwable
+     * @throws Throwable
      */
     public function render($request, Throwable $exception)
```
Comment (HerbertIV) [inline src/Http/Exceptions/Handler.php:20]: In such situations, try to use a strategy.

File: src/Http/Requests/FileListingRequest.php [hunk 2]
```diff
@@ -22,6 +23,31 @@ public function authorize()
...
+    public function getPage(): int
+    {
+        $page = $this->get('page');
+        return $page == null ? 1 : $page;
+    }
+
+    public function getPerPage(): int
+    {
+        $perPage = $this->get('perPage');
+        return $perPage == null ? 50 : $perPage;
```
Comment (HerbertIV) [inline src/Http/Requests/FileListingRequest.php:46]: You will get the same thing if you use the default value: `return $this->get('perPage', 50);`

File: src/Http/Requests/FileListingRequest.php [hunk 3]
```diff
@@ -22,6 +23,31 @@ public function authorize()
...
+
+    public function getDirectory(): string
+    {
+        return $this->get('directory');
+    }
+
+    public function getPage(): int
+    {
+        $page = $this->get('page');
+        return $page == null ? 1 : $page;
```
Comment (HerbertIV) [inline src/Http/Requests/FileListingRequest.php:40]: You will get the same thing if you use the default value: `return $this->get('page', 1);`

File: tests/Api/FilesApiListTest.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        ]);
+    }
+
+    public function testDirectoryListGivenDirectory()
+    {
+        $file1 = UploadedFile::fake()->image('test.png');
+        $file2 = UploadedFile::fake()->create('test.txt', 3, 'text/plain');
+
+        $directory = '/test';
+        $path = rtrim('/storage'.$directory,'/');
```
Comment (HerbertIV) [inline tests/Api/FilesApiListTest.php:51]: Why are you using 'rtrim' here when you set the value yourself?

File: src/Http/Exceptions/Handler.php (hunk 1 above)
Comment (qunabu) [reply src/Http/Exceptions/Handler.php:20]: IMHO this file can be removed `src/Http/Exceptions/Handler.php` 

Isn't is enought just to `throw CustomException` ? @HerbertIV

File: src/Http/Requests/FileListingRequest.php (hunk 3 above)
Comment (qunabu) [inline src/Http/Requests/FileListingRequest.php:40]:
```suggestion
        return $this->get('page', 1);
```

File: tests/TestCase.php [hunk 5]
```diff
@@ -3,18 +3,32 @@
...
     }
+
+    protected function getWithQuery(string $url, array $parameters, array $headers = []): TestResponse
+    {
+        if (empty($parameters)) {
+            $query = $url;
+        } else {
+            $query = $url.'?'.http_build_query($parameters);
+        }
+        return $this->get($query, $headers);
```
Comment (HerbertIV) [inline tests/TestCase.php:32]:
```suggestion
        return $this->get(
        rtrim(implode('?', [$url, http_build_query($parameters)]), '?')
        , $headers);
```

**Review by HerbertIV — CHANGES_REQUESTED** [review]
It would be nice to use the strategies I mentioned

File: src/Http/Exceptions/Handler.php (hunk 1 above)
Comment (HerbertIV) [reply src/Http/Exceptions/Handler.php:20]: Yes CustomException is fine, but it's worth sending the error status and content for that Exception when raising the exception.

File: tests/TestCase.php (hunk 5 above)
Comment (qunabu) [inline tests/TestCase.php:32]: or 
```suggestion
        return $this->get(empty($parameters) ? $url : , $url.'?'.http_build_query($parameters));
```

Comment (shdpl) [reply tests/TestCase.php:32]: isn't original version more readable?

File: src/Http/Exceptions/Handler.php (hunk 1 above)
Comment (shdpl) [reply src/Http/Exceptions/Handler.php:21]: Please check my Renderable{} interface solution.

File: tests/TestCase.php (hunk 5 above)
Comment (qunabu) [reply tests/TestCase.php]: For me thats not an issue at all

**Pull request merged**